rewrite-java-21/src/main/java/org/openrewrite/java/isolated/ReloadableJava21ParserVisitor.java

```diff
@@ -877,9 +912,26 @@ public J visitLambdaExpression(LambdaExpressionTree node, Space fmt) {
 
     @Override
     public J visitLiteral(LiteralTree node, Space fmt) {
-        cursor(endPos(node));
+        int endPos = endPos(node);
         Object value = node.getValue();
-        String valueSource = source.substring(((JCLiteral) node).getStartPosition(), endPos(node));
+
+        if (endPos == Position.NOPOS) {
+            if (typeMapping.primitive(((JCLiteral) node).typetag) == JavaType.Primitive.String) {
+                int quote = source.substring(cursor).startsWith("\"\"\"") ? 3 : 1;
+                if (quote == 3) {
+                    endPos = cursor + quote + source.indexOf("\"\"\"", cursor + quote) + quote;
+                } else {
+                    endPos = cursor + quote + value.toString().length() + quote;
+                }
+            } else {
+                endPos = cursor + indexOf(source.substring(cursor),
+                        ch -> Character.isWhitespace(ch) || ",;)]}+-*/%=!<>&|^?:.".indexOf(ch) != -1
```

Comment:
If we want to use this `indexOf()` utility, I think we should add an overload that takes an additional `int fromIndex` parameter, so that we can get rid of this `substring()` call.